src/hexkit/log.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Configurable logging tools with JSON output."""
+
+import json
+from collections import OrderedDict
+from datetime import datetime, timezone
+from logging import (
+    Formatter,
+    Logger,
+    LogRecord,
+    StreamHandler,
+    addLevelName,
+)
+from typing import Literal
+
+from pydantic import Field
+from pydantic_settings import BaseSettings
+
+from hexkit.correlation import (
+    correlation_id_var,
+)
+
+__all__ = [
+    "JsonFormatter",
+    "LoggingConfig",
+    "RecordCompiler",
+    "configure_logging",
+]
+
+
+DUMMY_RECORD = LogRecord("dummy", 0, "dummy", 0, None, None, None, None, None)
+RESERVED_RECORD_KEYS = DUMMY_RECORD.__dict__.keys()
+
+# Add TRACE log level
+addLevelName(5, "TRACE")
+LogLevel = Literal["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG", "TRACE"]
+
+
+class LoggingConfig(BaseSettings):
+    """A class containing logging config."""
+
+    log_level: LogLevel = Field(
+        default="INFO", description="The minimum log level to capture."
+    )
+    service_name: str = Field(
+        ...,
+        examples=["my-cool-special-service"],
+        description="The name of the (micro-)service. This will be included in log messages.",
+    )
+    service_instance_id: str = Field(
+        ...,
+        examples=["germany-bw-instance-001"],
+        description=(
+            "A string that uniquely identifies this instance across all instances of"
+            + " this service. This is included in log messages."
+        ),
+    )
+    log_format: str = Field(
+        default="",
```

Comment:
```suggestion
    log_format: Optional[str] = Field(
        None,
```

Reply:
Better keep the `default=` since otherwise mypy and pyright have problems.

Reply:
Thanks, I've made the change in [95ad845](https://github.com/ghga-de/hexkit/pull/77/commits/95ad845d31aed82970ecc469a48c74ea70978a29)

Reply:
OK, didn't knew that this is causing issues.

Reply:
The main issue is that you get errors when you're instantiation the model without specifying that field. We often silence these with comments but it's not necessary if you define the field with the default keyword.